### ISSUE TITLE
Duplicate assertion error being thrown during IdP-initiated response after sp logout

### ISSUE THREAD

**Issue by aas395** [simplesamlphp/simplesamlphp#237]
Hi,

We recently fixed the duplicate assertion issue that we were seeing in #230, and there is just one more case where we're seeing this where we think we shouldn't be.

If you log out from your session using lib/SimpleSAML/Auth/Simple::logout and resend the same assertion from the same open session in an IdP-first flow, I'm seeing a duplicate assertion error that looks like it's being thrown somewhere around line 124 depending on which version of the library you're using. Here's the code on that line in my version:

$e = new SimpleSAML_Error_Exception('Received duplicate assertion.');

Note that this is after implementing the patch proposed in issue #230.

I don't have the ability to test this myself right now, annoyingly, but it seems that after the logout the check that was taking place to make sure that the assertion was coming from the right place isn't happening correctly anymore since the session has been wiped.

Any thoughts about what we might be able to do to fix this? Thank you so much for your help.

**Comment by aas395** [comment]
Any thoughts on this?

**Comment by jaimeperez** [comment]
Hi @aas395!

At first sight, your scenario sounds highly unlikely (at least by accident). How are you seeing this? What's the procedure that's leading to this weird situation?

The problem is... when you get a duplicate assertion after being successfully logged in, it is completely ok to just ignore it because it is likely a mistake of the user, and you are already logged in, so you can just continue doing what you were doing. If you've been logged out, though, I see no way this situation could happen by accident. You are no longer logged in (and you don't have a way to tell if you *were* before, since your session is gone), so you can't just ignore the assertion. But at the same time, you of course can't honour it and log the user in, since you are sure it's a duplicate because you have it already in the store. So basically you need to fail horribly because something nasty is going on.

It would be nice to know a bit more about this, what's the situation leading to the error, and what's the reason for it (basically to see if there's a legit succession of events), but in the end I don't see a reasonable alternative to an error. We could change the message, of course, but I don't think that's going to help.

I'm closing the issue for now, but please, add any comments you may have and feel free to ask for it to be reopened if you think I'm wrong!

**Comment by aas395** [comment]
Hi Jamie,

You're right that this is obscure, but one of our partners is asking us to resolve this.

Here's the simplified flow to see this issue:

1) User authenticates using the IdP-first flow to our platform. The window where they started the SSO process stays open while a new window opens to handle the SSO assertion to our platform.
2) User does some things on our side and then logs out of our platform (which does not log the user out of the IdP).
3) User goes back to the IdP in the original browser tab/window (without refreshing that page) and sends the original SSO request to our side. They trigger a "duplicate assertion" error.

I'm hoping for a solution for this case or an explanation about why this is a security issue and we can't handle this situation differently. The other third parties this partner works with reportedly handle this re-request without throwing an error, so I'm hoping there's a way to do it here.

Thank you so much for your help!

**Comment by jaimeperez** [comment]
Hi again!

Let me first try to clarify something. When you say *other third parties this partner works with reportedly handle this re-request without throwing an error*, does that mean they just log the user in again, ignoring the fact that the assertion is a duplicate?

**Comment by aas395** [comment]
Hi Jaime,

I checked in with our partner, and one of their vendors said this about this situation:

"In your situation – it would be a valid request.  As long as the person is logged into the main application (the IdP), any request they make from that application is valid.  This is how SSO should work.  So anytime someone is successfully in the IdP and tries to access an SP it will let them.  However, if they sign-out of both, they would not be able to get back into the SP without first logging into the IdP.  Any logout or timeout rules to control what happens in the IdP should be enforced there, vendors should only have to worry about someone who is logged in then accessing their site.  We cannot create logic on our side that accounts for if they have been in IdP for too long, etc."

So it sounds like they support this, and from what I understand about SSO (particularly in the IdP-initiated flow), this sounds like the right approach. I'll ask specifically about the duplicate assertion checks that they do, but I wanted to share this with you in the mean time.

Thank you for your help!

**Comment by jaimeperez** [comment]
Hi again! And thanks for sharing this with us! :smile: 

I'm afraid this vendor either didn't understand the problem, or they are terribly confused. What they are describing is normal SSO, and they are absolutely correct in everything they say, but that's not the problem at hand. The issue here is not related to timeouts nor to normal SSO. SSO is not involved here for two reasons:

* SLO happened already, so there is **no valid session** in the IdP. When you logout in the SP, you trigger SLO, effectively logging out from any other SP you are logged into, and also from the IdP.
* The response received is a duplicate. It's not a normal IdP-initiated response, nor a response to a legitimate request, it's a duplicate. Duplicate assertions **must be discarded always**. Period. There can't be any discussion about that. 

Now the question is how to deal with the case, if you want to make the situation explicit to the user or not, and even what you can do in that position. As I said before, there's not much that we can do at that point, because the session has ended. We cannot accept the response and log the user in again, and if we use the RelayState or any other configured URL to redirect the user to some other page, he or she will not be logged in and that could fail, so I don't see much room for improvement, unfortunately... :disappointed:

**Comment by aas395** [comment]
Hi Jaime,

Thank you for this. Is it a requirement that we use SLO? When we log someone out of our SP, it doesn't log the user out of the IdP. I don't think it a log out of our system should also trigger an IdP logout.

If SLO isn't a requirement, I think that changes what you're describing. There is still a valid session on the IdP, so it's similar to #230 except that we have ended our existing SP session.

Let me know what you think.

**Comment by jaimeperez** [comment]
Hi!

Well... In a SSO scenario, there can't be normal logout, there can only be SLO. It's not that SLO is a requirement (that's not something SSP can impose, that's up to the federation's policies). But consider what happens when you perform a local logout in your SP, without propagating it as a proper SLO. As soon as the user clicks *login* again, he or she will be redirected again to the IdP, and will return back with a fresh, valid SAML response. The result: the user is logged in again in your SP, and your local logout is providing nothing but a false sense of security.

Of course you are free to trigger or not a SLO from your SP. But if you are not triggering a SLO, you shouldn't log users out locally either. So either you don't have logout at all (in which case the SP would disregard the duplicate assertion, as in #230), or you have full SLO, and the duplicate assertion is discarded, not only because it's a duplicate, but also because the session in the IdP is over and therefore it shouldn't be considered valid.

The only thing I could imagine in this case is having some URL that triggers a login from the SP, then telling the SP *by means of the configuration* that it should redirect there if a duplicate assertion is received when no local session is available. That way, authentication starts over and the user will end up in the IdP, where he or she will be able to decide whether to login again or not.

From your quotation above it's unclear to me whether that other vendor is accepting duplicates or not to log the user in again after a logout, but just in case: **duplicate assertions should never be used**. If they are blindly consuming duplicate assertions, they are doing it wrong and they have a serious security issue.

**Comment by aas395** [comment]
So in our case, users will have the ability to log in either through the SP's login page or through the IdP. I don't think our sign-out should trigger a sign-out on the IdP side. Users will need to be able to log out from our side, but even if we didn't give them the ability to log out on our side, we would still have differences in session lifetimes between the IdP and the SP. So our SP session might expire before the IdP session expired, which would bring us to the same problem where they attempt to log in to the SP from the still-valid IdP session by sending a duplicate assertion.

I don't know if our IdP has any sort of public URL where we can re-authorize SSO sessions. Since they use an IdP-first flow, I don't know that they would need to provide anything like that. I agree that duplicate assertions should be discarded, and I like your suggestion about adding a configurable option to make an attempt to re-authorize the session when duplicate assertions are received.

What do IdPs who use an IdP-first flow typically do in situations like this? I'm still new to the SSO world and I haven't seen this implemented anywhere before this project.

Thank you again for your patience and your help.

**Comment by jaimeperez** [comment]
No problem, I'm glad to help with whatever I can :smiley: 

Actually, your SP should trigger SLO. As I said, a local-only logout in an SSO environment makes no sense (it basically does nothing), and can be counter-productive if it contributes to give the user a false sense of security. Some IdPs allow the user to decide whether he or she wants to logout from all services, or just from the service that initiated SLO. That's what we do in Feide, for instance, and I think it's a feature in SimpleSAMLphp. That way, you are being explicit, and therefore there cannot be any confusions or misunderstandings. The user is conscious that sessions at other services are still alive if he or she clicks the button to do a local-only logout.

Timeouts for sessions have either nothing to do with this particular scenario. It is completely normal that SPs have different session lengths than IdPs. An SP should consider (and in most cases honour) the session length given by the IdP. In any case, that's the SP choice, and in some cases it might make sense to keep a longer session (i.e. apps in a mobile device that's only used by one person). If the SP session expires first, that's the usual scenario. You just ask for authentication again, and if the IdP has a valid session still going, SSO will happen and the user will be logged in seamlessly.

Now, I think the problem here is down to this:

> So our SP session might expire before the IdP session expired, which would bring us to the same problem where they attempt to log in to the SP from the still-valid IdP session by sending a duplicate assertion.

If the procedure (or flow, or whatever we want to call it) set by the IdP to login to services leads to sending duplicate assertions, then the procedure is wrong. That could be an unfortunate, rare accident, but it cannot be the norm. If this is happening often, they need to review their procedure.

I wonder why they are imposing an IdP-first flow. Are they supporting SAML 1.1 only? The IdP-first is nice to have in some situations (for example, when you want to provide links to services from a centralised portal), but it stops being a good idea if that's **the only flow** you support. In normal circumstances, when your SP session expires, and the user tries to login again, you just send another SAML request to the IdP and if there's a valid session in the IdP, the user is logged in again. In other words, IdP-first is good when it's complimentary to the normal SAML flow, not when it is exclusive. If this is due to supporting SAML 1.1 only, SAML 2.0 is more than 10 years old now :wink: 

In the end, I think the source of all this trouble is the fact that they are submitting the SAML response to you in a new window, keeping the original open. That's always a very bad idea, and now you are experiencing why. If the window keeps open, users are mislead to think they can just go back there and re-submit to login again. Which they can't.

If they want their users to stop seeing *duplicate assertion* errors in your SP, all they need to do is to fix their IdP to **not** open new windows in the middle of a SAML flow. They will probably realize then of how painful is the IdP-first-only flow for their users, and hopefully they'll consider fixing that too. Or maybe not and their users will be more frustrated, but they will see no more *duplicate assertion* errors. Unless users are so used to this workflow that they start *forcing* somehow the SAML flow to continue in a separate window, but that will again be their own fault.

I know in your position the customer is always right, but just because they decided to do something wrong, you shouldn't do something *terribly wrong* (i.e. accepting duplicate assertions).

**Comment by aas395** [comment]
This is incredibly helpful. I appreciate your thoroughness and your patience as I figure this out.

So as I prepare to take this back to our partners, I'm trying to come up with a simplified way to explain this. Am I right in thinking that there can only be one valid SAML assertion per SP session, and that once that session on the SP is destroyed, the same SAML assertion will no longer be valid?

If that's the case, it's clear why new-window assertions are a bad idea. It's also clear that they would want to allow us a way to log users in and out without needing to start on the IdP.

Providing single log-in and log-out endpoints for us would be nice, although I still don't think that would fix our particular issue if they're sending duplicate assertions and we're always discarding them. Does SimpleSamlPHP give us the option to return to the IdP SSO login page if we receive a duplicate assertion rather than throwing an error? I just need to be able to articulate what our options are.

**Comment by jaimeperez** [comment]
Hi!

It's more or less like that, but not exactly. The session at the SP is orthogonal to how you use assertions. As a rule of thumb, you will only need a new assertion when you don't have a session or it is expired or cancelled, but the main thing is that the IdP knows nothing about the session in the SP, so it will just issue assertions on demand.

The reason why one single assertion cannot be duplicated is security. If an attacker manages to grab an assertion while you are logging in, a replay attack could be mounted to obtain access. That means if an SP does not check for assertion duplicates and let them through, the attacker can send again an assertion that has been previously used, and get logged in as the victim. That's why I've been so stubborn regarding duplicate assertions.

So that's what I would tell your customer: duplicate assertions cannot be used because of security reasons, as allowing that would create a serious vulnerability in the service. Sending the assertions in a different window allows your users to send duplicate assertions, leading to an error. If you want to get rid of the error, you need to stop opening new browser windows.

Regarding your final question, I don't think there's a way to do that today. It could be implemented, though, but that would be a bit more complicated than solving the issue your reported in #230.

**Comment by aas395** [comment]
Got it. The essence of this is basically "don't send assertions in a new window and you won't have problems".

I don't personally need the configuration change I suggested earlier -- I took all this feedback to our partner and they'll be making changes to the flow to prevent this situation (still sending assertions in a new window but also reloading the current page to prepare a new assertion).

Thank you again for all your help!
